### PR TITLE
[KBSWITCH][CPL:INPUT][NTUSER][EXPLORER] Fix keyboard layout icon

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Fix keyboard layout icon in taskbar notification area.
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700), [CORE-2699](https://jira.reactos.org/browse/CORE-2699), [CORE-18546](https://jira.reactos.org/browse/CORE-18546)

## Proposed changes

- Call `ActivateKeyboardLayout` to select the keyboard layout correctly.
- Modify `WM_INPUTLANGCHANGEREQUEST` parameter.
- Modify `BroadcastSystemMessageW` parameter.
- Revert Taskbar Notification Area `MA_NOACTIVATE` HACK https://github.com/reactos/reactos/commit/83442919e4df0067a7bc72f351fefed667c1737a . This fixes Context Menu display.
- Load the `"IME File"` value and set the IME icon if necessary.
- Correctly implement global hooks.

## TODO

- [x] `Alt+Shift` on Command Prompt (HACK).
- [x] `Alt+Shift` on Notepad on Win2k3 and ReactOS with `kbswitch`.
- [x] `[半/全]` key on Japanese ReactOS (with IME HKL defaulted).
- [x] Do tests.

### CODE REVIEW

File: dll/cpl/input/input_list.c [hunk 1]
```diff
@@ -447,15 +447,28 @@ InputList_Process(VOID)
...
     if (SystemParametersInfoW(SPI_SETDEFAULTINPUTLANG, 0, &pCurrent->hkl, 0))
     {
-        DWORD dwRecipients = BSM_ALLCOMPONENTS | BSM_ALLDESKTOPS;
+        DWORD dwRecipients = BSM_ALLDESKTOPS | BSM_APPLICATIONS; /* confirmed in 2k and xp */
 
         BroadcastSystemMessageW(BSF_POSTMESSAGE,
                                 &dwRecipients,
                                 WM_INPUTLANGCHANGEREQUEST,
-                                0,
+                                1,
```
Comment (binarymaster) [inline dll/cpl/input/input_list.c:471]:
```suggestion
                                INPUTLANGCHANGE_SYSCHARSET,
```

File: base/applications/kbswitch/kbswitch.c [hunk 2]
```diff
@@ -289,7 +289,7 @@ GetLayoutIDByHkl(HKL hKl, LPTSTR szLayoutID, SIZE_T LayoutIDLength)
 static BOOL CALLBACK
 EnumWindowsProc(HWND hwnd, LPARAM lParam)
 {
-    PostMessage(hwnd, WM_INPUTLANGCHANGEREQUEST, 0, lParam);
+    PostMessage(hwnd, WM_INPUTLANGCHANGEREQUEST, 1, lParam);
```
Comment (binarymaster) [inline base/applications/kbswitch/kbswitch.c:292]:
```suggestion
    PostMessage(hwnd, WM_INPUTLANGCHANGEREQUEST, INPUTLANGCHANGE_SYSCHARSET, lParam);
```

**Review by binarymaster — COMMENTED** [review]
`0x0001` is `INPUTLANGCHANGE_SYSCHARSET` according to: https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-inputlangchangerequest

**Comment by katahiromz** [comment]
VirtualBox_Windows XP_01_01_2008_10_48_32
Looks good.

File: base/shell/explorer/trayntfy.cpp [hunk 3]
```diff
@@ -124,11 +124,6 @@ class CTrayNotifyWnd :
         return TRUE;
     }
 
-    LRESULT OnMouseActivate(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
-    {
-        return MA_NOACTIVATE;
```
Comment (katahiromz) [inline base/shell/explorer/trayntfy.cpp:129]: This was HACK.

File: dll/cpl/input/input_list.c [hunk 4]
```diff
@@ -450,12 +453,12 @@ InputList_Process(VOID)
     /* Change the default keyboard language */
     if (SystemParametersInfoW(SPI_SETDEFAULTINPUTLANG, 0, &pCurrent->hkl, 0))
     {
-        DWORD dwRecipients = BSM_ALLCOMPONENTS | BSM_ALLDESKTOPS;
+        DWORD dwRecipients = BSM_ALLDESKTOPS | BSM_APPLICATIONS; /* confirmed in 2k and xp */
```
Comment (HBelusca) [inline dll/cpl/input/input_list.c:456]: Do we need this comment?

File: base/shell/explorer/trayntfy.cpp (hunk 3 above)
Comment (HBelusca) [reply base/shell/explorer/trayntfy.cpp:129]: Please say in your commit log that this reverts commit 83442919e :
https://github.com/reactos/reactos/commit/83442919e4df0067a7bc72f351fefed667c1737a

File: base/applications/kbswitch/kbswitch.c [hunk 5]
```diff
@@ -654,10 +751,25 @@ _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInst, LPTSTR lpCmdLine, INT nCmdSh
...
+        {
+            if (!GetMessage(&msg, NULL, 0, 0))
+                break;
+
+            TranslateMessage(&msg);
+            DispatchMessage(&msg);
+        }
+        else
+        {
+            if (GetTickCount() - dwOldTick >= 700)
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:769]: Why not using a standard WM_TIMER message firing every 700 ms in your WndProc ?

File: base/applications/kbswitch/kbswitch.c [hunk 6]
```diff
@@ -465,6 +556,8 @@ WndProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
     static UINT s_uTaskbarRestart;
     POINT pt;
     HMENU hLeftPopupMenu;
+    LRESULT ret = 0;
+    HWND hwndFore;
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:563]:
```suggestion
```

File: base/applications/kbswitch/kbswitch.c [hunk 7]
```diff
@@ -495,7 +589,9 @@ WndProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 
         case WM_WINDOW_ACTIVATE:
         {
-            return UpdateLanguageDisplayCurrent(hwnd, wParam);
+            hwndFore = GetForegroundWindow();
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:595]:
```suggestion
            HWND hwndFore = GetForegroundWindow();
```

File: base/applications/kbswitch/kbswitch.c [hunk 8]
```diff
@@ -131,18 +142,82 @@ GetLayoutName(LPCTSTR szLayoutNum, LPTSTR szName, SIZE_T NameLength)
...
+static BOOL CALLBACK
+EnumResNameProc(
+    HMODULE hModule,
+    LPCTSTR lpszType,
+    LPTSTR lpszName,
+    LPARAM lParam)
+{
+    PLOAD_ICON pLoadIcon = (PLOAD_ICON)lParam;
+    pLoadIcon->hIcon = (HICON)LoadImage(hModule, lpszName, IMAGE_ICON,
+                                        pLoadIcon->cxIcon, pLoadIcon->cyIcon, 0);
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:204]:
```suggestion
                                        pLoadIcon->cxIcon, pLoadIcon->cyIcon,
                                        LR_DEFAULTCOLOR);
```

**Comment by katahiromz** [comment]
https://stackoverflow.com/questions/70979849/windows-10-detect-keyboard-layout-change

**Comment by HBelusca** [comment]
It can be of interest to look at some of those functions I pointed out yesterday on mattermost, or study how internat.exe (on Win2k) or msctf.dll / msutb.dll (on WinXP/2k3) communicates with win32k ?

**Comment by katahiromz** [comment]
The PR should work now.

**Comment by katahiromz** [comment]
I found a problem on Command Prompt:
VirtualBox_ReactOS_27_10_2022_10_05_31
Cannot change the keyboard layout.

**Comment by katahiromz** [comment]
WinXP `ActivateKeyboardLayout` is not compatible to WinNT. There's process boundary and restricted not to activate from the outer process.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_27_10_2022_20_40_04
Corrected! Let's go!

**Review by HBelusca — APPROVED** [review]
Regarding the "FIXME console", I suppose it's also because it is hosted in the special CSRSS process (which we know has some special handling in Win32k as well)

File: base/applications/kbswitch/kbswitch.c [hunk 9]
```diff
@@ -450,12 +581,76 @@ UpdateLanguageDisplay(HWND hwnd, HKL hKl)
...
+
+    if (lstrcmpi(szClass, szKbSwitcherName) == 0 ||
+        lstrcmpi(szClass, TEXT("Shell_TrayWnd")) == 0)
+    {
+        return FALSE; /* Special window */
+    }
+
+    /* FIXME: CONWND is multithreaded but KLF_SETFORPROCESS and
+              DefWindowProc.WM_INPUTLANGCHANGEREQUEST won't work yet */
+    if (lstrcmpi(szClass, TEXT("ConsoleWindowClass")) == 0)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply base/applications/kbswitch/kbswitch.c]: Used `_tcsicmp` instead of `_tcscmp`.

**Review by HBelusca — APPROVED** [review]

**Pull request merged**